src/box/vy_upsert.c

```diff
@@ -132,6 +132,17 @@ vy_apply_upsert_on_terminal_stmt(struct tuple *upsert, struct tuple *stmt,
 			ups_ops = ups_ops_end;
 			continue;
 		}
+		/*
+		 * Result statement must satisfy space's format. Since upsert's
+		 * tuple correctness is already checked in vy_upsert(), let's
+		 * use its format to provide result verification.
+		 */
+		struct tuple_format *format = tuple_format(upsert);
+		if (tuple_validate_raw(format, exec_res) != 0) {
+			if (!suppress_error)
+				diag_log();
+			continue;
```

Comment:
Looks like there is another bug unrelated to the one being fixed. Should not we have `ups_ops = ups_ops_end;` here too?

Reply:
Good catch! Fixed and added a test.

<details><summary>diff</summary>

```diff
diff --git a/src/box/vy_upsert.c b/src/box/vy_upsert.c
index 9b18f099cedb..604184391363 100644
--- a/src/box/vy_upsert.c
+++ b/src/box/vy_upsert.c
@@ -141,6 +141,7 @@ vy_apply_upsert_on_terminal_stmt(struct tuple *upsert, struct tuple *stmt,
 		if (tuple_validate_raw(format, exec_res) != 0) {
 			if (!suppress_error)
 				diag_log();
+			ups_ops = ups_ops_end;
 			continue;
 		}
 		/*
diff --git a/test/vinyl-luatest/gh_10099_invalid_upsert_test.lua b/test/vinyl-luatest/gh_10099_invalid_upsert_test.lua
index 7a11b48139a6..2349f7bbb78d 100644
--- a/test/vinyl-luatest/gh_10099_invalid_upsert_test.lua
+++ b/test/vinyl-luatest/gh_10099_invalid_upsert_test.lua
@@ -26,9 +26,17 @@ g.test_invalid_upsert = function(cg)
         s:create_index('pk', {parts = {{1, 'integer'}}})
         s:insert({1})
         s:insert({2})
+        s:insert({3})
         box.snapshot()
+        box.begin()
         s:upsert({1}, {{'#', 1, 1}})
         s:upsert({2}, {{'=', 1, 's'}})
-        t.assert_equals(s:select({}, {fullscan = true}), {{1}, {2}})
+        s:upsert({3}, {{'=', 1, 1}})
+        s:upsert({1}, {{'!', 2, 10}})
+        s:upsert({2}, {{'!', 2, 20}})
+        s:upsert({3}, {{'!', 2, 30}})
+        box.commit()
+        t.assert_equals(s:select({}, {fullscan = true}),
+                        {{1, 10}, {2, 20}, {3, 30}})
     end)
 end
```

</details>